### PR TITLE
[LW] Retention LWEC on max size

### PR DESCRIPTION
**Pull request by Jolyon-S**
**Goals (and why)**:
* There are cases where it doesn't make sense to only retention on in-flight transactions. In particular, if you have two clients, and one of them is slower than the other, and the slow one talks to Atlas on the fast server, then it will receive a snapshot and cancel all in-flight transactions on the slow one.

**Implementation Description (bullets)**:
* Change retention strategy in `VersionedEventStore` to go on max size instead of living transactions
* This breaks the invariant that we always keep enough versions around - therefore, we need further checks that we have the right versions around.
* Some refactoring around how the checks are carried out improves clarity.

**Testing (What was existing testing like?  What have you done to improve it?)**:
There are several tests:
* TimestampStateStoreTest
* VersionedEventStoreTest
* LockWatchEventCacheIntegrationTest
Much of the behaviour stayed the same, apart from retention. Any retention-based tests were modified according to the new behaviour, and two tests were added.

**Concerns (what feedback would you like?)**:
* Correctness most of all - I want to make sure we're not shooting ourselves at any point here.
* Also, we want to make sure we're not making this inefficient (e.g. retrying too many transactions).

**Where should we start reviewing?**:
`VersionedEventStore` is probably the best place to start, then spread out from there. Other key classes are `LockWatchEventLog` and `LockWatchEventCacheImpl`.

**Priority (whenever / two weeks / yesterday)**:
We want to get this merged before end of day tomorrow.

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheImpl.java [hunk 1]
```diff
@@ -153,27 +144,40 @@ LockWatchEventCacheState getStateForTesting() {
     }
 
 
-    private void assertTrue(boolean condition, String message) {
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheImpl.java:156]: moved these around and made some static; the diff renders weirdly here

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/VersionedEventStore.java [hunk 2]
```diff
@@ -44,14 +51,15 @@
...
+    LockWatchEvents retentionEvents() {
+        int numToRetention = Math.max(0, eventMap.size() - maxEvents);
+        LockWatchEvents.Builder builder = new LockWatchEvents.Builder();
+        Iterators.consumingIterator(Iterators.limit(eventMap.entrySet().iterator(), numToRetention))
+                .forEachRemaining(entry -> builder.addEvents(entry.getValue()));
+        return builder.build();
     }
 
-    boolean contains(long key) {
+    boolean containsEntriesLessThanOrEqualTo(long key) {
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/VersionedEventStore.java:62]: drive-by rename for clarity

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java [hunk 3]
```diff
@@ -287,11 +304,13 @@ public void nonContiguousEventsThrows() {
...
     @Test
     public void upToDateClientDoesNotThrow() {
+        eventCache = createEventCache(6);
         setupInitialState();
         eventCache.processStartTransactionsUpdate(TIMESTAMPS_2, SUCCESS);
-        assertThat(eventCache.getUpdateForTransactions(
-                TIMESTAMPS,
-                Optional.of(LockWatchVersion.of(LEADER, SUCCESS_VERSION))).events()).isEmpty();
+        assertThatCode(() -> eventCache.getUpdateForTransactions(
+                ImmutableSet.of(START_TS, 16L),
```
Comment (jkozlowski) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java:311]: I'm assuming you just wanted to make this test do a bit more by adding the extra timestamp to lookup?

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java [hunk 4]
```diff
@@ -287,11 +304,13 @@ public void nonContiguousEventsThrows() {
...
+                Optional.of(LockWatchVersion.of(LEADER, 3L))))
```
Comment (jkozlowski) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java:312]: Hmm, I feel like I'd rather keep it as an assertion, and maybe rename the test? And if you add the "latestVersion" to TransactionUpdate that can tighten this test as well?

**Review by jkozlowski — COMMENTED** [review]
Needs a bit of rejigging.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ClientLockWatchSnapshot.java [hunk 5]
```diff
@@ -61,12 +62,16 @@ private ClientLockWatchSnapshot() {
     }
 
     void processEvents(LockWatchEvents events, UUID versionId) {
-        if (!events.latestSequence().isPresent()) {
+        if (events.events().isEmpty()) {
             return;
         }
 
+        Preconditions.checkArgument(events.versionRange().isPresent(),
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ClientLockWatchSnapshot.java:69]: Should be impossible, I just don't like doing `Optional.get()` blindly. I've moved to `LockWatchEvents`.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ClientLogEvents.java [hunk 6]
```diff
@@ -16,27 +16,25 @@
...
 
 @Value.Immutable
 interface ClientLogEvents {
 
-    List<LockWatchEvent> events();
+    LockWatchEvents events();
 
     boolean clearCache();
 
     default TransactionsLockWatchUpdate map(Map<Long, LockWatchVersion> timestampMap) {
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ClientLogEvents.java:34]: Makes sense. Done

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheImpl.java [hunk 7]
```diff
@@ -53,12 +50,18 @@
  * in concurrency issues and inconsistency in the cache state.
  */
 public final class LockWatchEventCacheImpl implements LockWatchEventCache {
+    /**
+     * This value should be at least as large as the max size in LockEventLogImpl. If it is smaller, it is possible that
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheImpl.java:54]: I'll reference the timelock class informally, but you can't link it here since this package does not know about it (I'd already tried this before).

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import com.palantir.lock.watch.LockWatchVersion;
+import com.palantir.logsafe.Preconditions;
+
+@Value.Immutable
+interface TimestampMapping {
+    Map<Long, LockWatchVersion> timestampMapping();
+
+    @Value.Derived
+    default Range<Long> versionRange() {
+        Optional<Long> firstVersion = timestampMapping()
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java:35]: That _is_ a cute one. Nice!

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Value.Derived
+    default UUID leader() {
+        Optional<UUID> leader = timestampMapping()
+                .values()
+                .stream()
+                .findAny()
+                .map(LockWatchVersion::id);
+
+        Preconditions.checkState(leader.isPresent(),
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java:49]: I'll extract the empty check out, but I really don't like doing unchecked `Optional.get()` calls!

[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java]: Actually, using `.distinct` makes this unnecessary anyway. Nice!

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+        Preconditions.checkState(firstVersion.isPresent() && lastVersion.isPresent(),
+                "Cannot compute timestamp mapping for empty map of timestamps");
+
+        return Range.closed(firstVersion.get(), lastVersion.get());
+    }
+
+    @Value.Derived
+    default UUID leader() {
+        Optional<UUID> leader = timestampMapping()
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java:43]: In any case, we don't want to return multiple leaders, because it will just cause those transactions to conflict further down the line anyway - so at best, this would be a retryable exception.

However, as you note, I don't think we should ever hit this, as a new leader will wipe all previous state, including current timestamp mappings (so we won't be able to get it from the cache - if timestamp versions are not present because of this kind of thing, we retry anyway).

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventLog.java [hunk 11]
```diff
@@ -133,29 +132,29 @@ private void processSuccess(LockWatchStateUpdate.Success success) {
         }
 
         if (success.lastKnownVersion() > latestVersion.get().version()) {
-            assertEventsAreContiguousAndNoEventsMissing(success.events());
+            assertNoEventsAreMissing(success.events());
             latestVersion = Optional.of(LockWatchVersion.of(success.logId(), eventStore.putAll(success.events())));
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventLog.java:136]: Not particularly - changed.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/VersionedEventStoreTest.java [hunk 12]
```diff
@@ -39,23 +40,24 @@
...
     @Test
-    public void getAndRemoveElementsUpToExclusiveDoesNotIncludeEndVersion() {
+    public void getAndRemoveElementsRemovesOldestElements() {
         eventStore.putAll(ImmutableList.of(EVENT_1, EVENT_2, EVENT_3));
-        LockWatchEvents events = eventStore.getAndRemoveElementsUpToExclusive(3L);
+        eventStore.putAll(ImmutableList.of(EVENT_4));
+        LockWatchEvents events = eventStore.retentionEvents();
         assertThat(events.events().stream().map(LockWatchEvent::sequence)).containsExactly(1L, 2L);
-        assertThat(events.latestSequence()).hasValue(2L);
+        assertThat(events.versionRange().map(Range::upperEndpoint)).hasValue(2L);
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/VersionedEventStoreTest.java:52]: That's true - I was just porting it over from the previous, but actually you're right in that it is redundant.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java [hunk 13]
```diff
@@ -116,7 +116,7 @@ boolean isDev() {
 
     @Before
     public void before() {
-        eventCache = new LockWatchEventCacheImpl(LockWatchEventLog.create());
+        eventCache = createEventCache(3);
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java:119]: Reasonable - done.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java [hunk 14]
```diff
@@ -240,7 +255,9 @@ public void getEventsForTransactionsReturnsSnapshotWithOldEvents() {
                 ImmutableMap.of(16L, LockWatchVersion.of(LEADER, 7L)));
         assertThat(results.events()).containsExactly(
                 LockWatchCreatedEvent.builder(ImmutableSet.of(REFERENCE),
-                        ImmutableSet.of(DESCRIPTOR, DESCRIPTOR_3)).build(6L),
+                        ImmutableSet.of(DESCRIPTOR, DESCRIPTOR_2)).build(4L),
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java:255]: 1. No compression
2. Retention change (i.e. keeps some values around).

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java (hunk 3 above)
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java:308]: Yes

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java (hunk 4 above)
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventCacheIntegrationTest.java:309]: I've actually changed the test:
- before, it was testing that a client that is on the exact same version as the cache will not throw
- now, it's testing that a client that is on the same version as the transaction (but _not_ the cache) will not throw.

It'd probably be best here if I make sure we have both of those tests.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ClientLogEvents.java [hunk 15]
```diff
@@ -16,27 +16,25 @@
...
         return ImmutableTransactionsLockWatchUpdate.builder()
                 .startTsToSequence(timestampMap)
-                .events(events())
+                .events(events().events())
```
[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ClientLogEvents.java:64]: Done

[the remark this answers is not in the document]
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/ClientLogEvents.java:64]: (somewhat - I didn't add latestVersion to TLWU).

**Review by jkozlowski — APPROVED** [review]
